File: beginner/maze/c++/maze.cpp
#include <queue>

using namespace std;

int bfs();

const int INF = 100000000;
const int MAX_N = 10;
const int MAX_M = 10;

// state class
typedef pair<int, int> P;

// ===== ↓↓↓input↓↓↓ ===== //
int N = 10; // vertical
int M = 10; // horizontal
int sx = 0, sy = 1; // start position
int gx = 9, gy = 8; // goal position
char maze[10][10] = {
		{'#', 'S', '#', '#', '#', '#', '#', '#', '.' , '#'},
		{'.', '.', '.', '.', '.', '.', '#', '.', '.' , '#'},
		{'.', '#', '.', '#', '#', '.', '#', '#', '.' , '#'},
		{'.', '#', '.', '.', '.', '.', '.', '.', '.' , '.'},
		{'#', '#', '.', '#', '#', '.', '#', '#', '#' , '#'},
		{'.', '.', '.', '.', '#', '.', '.', '.', '.' , '#'},
		{'.', '#', '#', '#', '#', '#', '#', '#', '.' , '#'},
		{'.', '.', '.', '.', '#', '.', '.', '.', '.' , '.'},
		{'.', '#', '#', '#', '#', '.', '#', '#', '#' , '.'},
		{'.', '.', '.', '.', '#', '.', '.', '.', 'G' , '#'}
	};
// ===== ↑↑↑input↑↑↑ ===== //

// shortest distance of per point
int d[MAX_N][MAX_N];

// move 4 direction vector
int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};

int main(int argc, char *argv[]) {
	int res = bfs();
    printf("%d\n", res);
}

// get shortest distance from (sx, sy) to (gx, gy)
// INF if you can not reach
int bfs() { // Breadth-first search : bfs
    queue<P> que;
    // init INF for all point
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            d[i][j] = INF;
        }
    }
    // set queue start position
    // this point distance zero
    que.push(P(sx, sy));
    d[sx][sy] = 0;

    // loop until the queue is empty
    while (que.size()) {
        // get queue of first
        P p = que.front();
        que.pop();
        printf("%d\n", que);
        
        // finish if state goal point
        if (p.first == gx && p.second == gy) break;
        
        // loop 4 derection   
        for (int i = 0; i < 4; i++) {
            // let (nx, ny) be the point after moving
            int nx = p.first + dx[i];
            int ny = p.second + dy[i];

            // judge move avaivale and exist visited
            if (0 <= nx && nx < N && 0 <= ny && ny < M && maze[nx][ny] != '#' && d[nx][ny] == INF) {
                // If you can move it, put it in the queue and determine the distance of that point as the distance from p + 1
                que.push(P(nx, ny));
                d[nx][ny] = d[p.first][p.second] + 1;
            }
        } 
    }
    return d[gx][gy];
}

File: beginner/breadth_search/maze/go/maze.go
package main

import (
	"fmt"
)

const INF = 100000000
const MAX_N int = 10
const MAX_M int = 10

type Pair struct {
	x, y int
}

var d = [MAX_N][MAX_M]int{}

var dx = [4]int{1, 0, -1, 0}
var dy = [4]int{0, 1, 0, -1}

// ===== ↓↓↓input↓↓↓ ===== //
var N, M, sx, sy, gx, gy int = 10, 10, 0, 1, 9, 8
var maze = [10][10]string{
	{"#", "S", "#", "#", "#", "#", "#", "#", ".", "#"},
	{".", ".", ".", ".", ".", ".", "#", ".", ".", "#"},
	{".", "#", ".", "#", "#", ".", "#", "#", ".", "#"},
	{".", "#", ".", ".", ".", ".", ".", ".", ".", "."},
	{"#", "#", ".", "#", "#", ".", "#", "#", "#", "#"},
	{".", ".", ".", ".", "#", ".", ".", ".", ".", "#"},
	{".", "#", "#", "#", "#", "#", "#", "#", ".", "#"},
	{".", ".", ".", ".", "#", ".", ".", ".", ".", "."},
	{".", "#", "#", "#", "#", ".", "#", "#", "#", "."},
	{".", ".", ".", ".", "#", ".", ".", ".", "G", "#"},
}

// ===== ↑↑↑input↑↑↑ ===== //

func main() {
	res := bfs()
	fmt.Println(res)
}

func bfs() int {
	que := make([]Pair, 0)
	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			d[i][j] = INF
		}
	}

	que = append(que, Pair{sx, sy})
	d[sx][sy] = 0

	for len(que) > 0 {
		p := que[0]
		que = que[1:]
		if p.x == gx && p.y == gy {
			break
		}

		for i := 0; i < 4; i++ {
			nx := p.x + dx[i]
			ny := p.y + dy[i]

			if 0 <= nx && nx < N && 0 <= ny && ny < M && maze[nx][ny] != "#" && d[nx][ny] == INF {
				que = append(que, Pair{nx, ny})
				d[nx][ny] = d[p.x][p.y] + 1
			}
		}
	}
	return d[gx][gy]
}
